Source/Initialization/ExternalField.cpp

```diff
@@ -189,99 +190,368 @@ ExternalFieldParams::ExternalFieldParams(const amrex::ParmParse& pp_warpx)
     //___________________________________________________________________________
 }
 
-ExternalFieldReader::ExternalFieldReader (std::string const& read_fields_from_path,
-                                          std::string const& F_name,
-                                          std::string const& F_component)
+ExternalFieldReader::ExternalFieldReader (
+    std::string read_fields_from_path,
+    std::string F_name, std::string F_component,
+    amrex::GpuArray<amrex::Real,AMREX_SPACEDIM> const& problo,
+    amrex::GpuArray<amrex::Real,AMREX_SPACEDIM> const& pdx,
+    amrex::Box const& pbox, bool distributed)
+    : m_file(std::move(read_fields_from_path)),
+      m_name(std::move(F_name)),
+      m_component(std::move(F_component)),
+      m_problo(problo),
+      m_probdx(pdx),
+      m_distributed(distributed)
+{
+    amrex::RealBox rbox;
+    for (int idim = 0; idim < AMREX_SPACEDIM; ++idim) {
+        rbox.setLo(idim, m_problo[idim] + pbox.smallEnd(idim)*m_probdx[idim]);
+        rbox.setHi(idim, m_problo[idim] + pbox.  bigEnd(idim)*m_probdx[idim]);
+    }
+
+    load_data(rbox);
+
+}
+
+void ExternalFieldReader::load_data (amrex::RealBox const& pbox)
 {
 #if defined(WARPX_USE_OPENPMD) && !defined(WARPX_DIM_RCYLINDER) && !defined(WARPX_DIM_RSPHERE)
+    using namespace amrex;
 
-    auto series = openPMD::Series(read_fields_from_path, openPMD::Access::READ_ONLY);
+    auto series = openPMD::Series(m_file, openPMD::Access::READ_ONLY);
     auto iseries = series.iterations.begin()->second;
-    auto F = iseries.meshes[F_name];
+    auto F = iseries.meshes[m_name];
 
-#if (AMREX_SPACEDIM > 1)
-    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(F.getAttribute("dataOrder").get<std::string>() == "C",
-                                     "Reading from files with non-C dataOrder is not implemented");
-#endif
+    bool c_order = F.getAttribute("dataOrder").get<std::string>() == "C";
+    amrex::ignore_unused(c_order);
 
     auto axisLabels = F.getAttribute("axisLabels").get<std::vector<std::string>>();
     auto fileGeom = F.getAttribute("geometry").get<std::string>();
 
+    bool xyz_order = true;
+
 #if defined(WARPX_DIM_3D)
     WARPX_ALWAYS_ASSERT_WITH_MESSAGE(fileGeom == "cartesian", "3D can only read from files with cartesian geometry");
-    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(axisLabels.at(0) == "x" && axisLabels.at(1) == "y" && axisLabels.at(2) == "z",
-                                     "3D expects axisLabels {x, y, z}");
+    if (axisLabels.at(0) == "x" && axisLabels.at(1) == "y" && axisLabels.at(2) == "z") {
+        xyz_order = true;
```

Comment:
The logic to add here would be:
```suggestion
        xyz_order = dataOrder == "C"? true : false;
```

Reply:
I don't think this is correct. See my comments above.